contracts/extensions/DeltaNeutralBasisTradingStrategyExtension.sol

```diff
@@ -0,0 +1,1456 @@
+                address(strategy.spotAssetAddress),
+                _baseUnits,
+                exchange.buyExactSpotTradeData      // buy exact amount
+            )
+            : abi.encodeWithSelector(
+                ITradeModule.trade.selector,
+                address(strategy.setToken),
+                exchange.exchangeName,
+                address(strategy.spotAssetAddress),
+                _baseUnits,
+                address(collateralToken),
+                _usdcUnits,
+                exchange.sellExactSpotTradeData     // sell exact amount
+            );
+
+        invokeManager(address(strategy.tradeModule), dexTradeCallData);
+    }
+
+    /**
+     * Invokes PerpV2BasisTradingModule to withdraw funding to be reinvested. Pass MAX_UINT_256 to withdraw all funding.
+     */
+    function _withdrawFunding(uint256 _fundingNotional) internal {
+        bytes memory withdrawCallData = abi.encodeWithSelector(
+            IPerpV2BasisTradingModule.withdrawFundingAndAccrueFees.selector,
+            strategy.setToken,
+            _fundingNotional
+        );
+
+        invokeManager(address(strategy.basisTradingModule), withdrawCallData);
+    }
+
+    /**
+     * Reinvests 50% of USDC position to spot asset and deposits the rest to PerpV2 to increase the short perp position.
+     * Used in the reinvest() function.
+     *
+     * return uint256           Calculated amount of funding to be reinvested
+     * return uint256           Spot amount increase notional. Returned on swapping 50% of funding to be reinvested
+     */
+    function _handleReinvest(LeverageInfo memory _leverageInfo) internal returns (uint256, uint256) {
+
+        uint256 defaultUsdcUnits = strategy.setToken.getDefaultPositionRealUnit(address(collateralToken)).toUint256();
+
+        if (defaultUsdcUnits == 0) { return (0, 0); }
+
+        uint256 setTotalSupply = strategy.setToken.totalSupply();
+        uint256 fundingReinvestedNotional = defaultUsdcUnits.preciseMul(setTotalSupply);  // 50%
+        uint256 spotAmountOutNotional = strategy.quoter.quoteExactInput(exchange.buySpotQuoteExactInputPath, fundingReinvestedNotional.div(2));
+        uint256 baseUnits = spotAmountOutNotional.preciseDiv(setTotalSupply);
+
+        // Increase spot position
+        _executeDexTrade(baseUnits, defaultUsdcUnits, true);
+
+        // Deposit rest
+        defaultUsdcUnits = strategy.setToken.getDefaultPositionRealUnit(address(collateralToken)).toUint256();
+        _deposit(defaultUsdcUnits);
+
+        // Increase perp position
+        _executePerpTrade(baseUnits.toInt256().neg(), _leverageInfo);
+
+        return (fundingReinvestedNotional, spotAmountOutNotional);
+    }
+
+    /* ============ Calculation functions ============ */
+
+    /**
+     * Calculate the current leverage ratio.
+     *
+     * return int256            Current leverage ratio
+     */
+    function _calculateCurrentLeverageRatio(ActionInfo memory _actionInfo) internal pure returns(int256) {
```

Comment:
Just for more information ... here we continue to measure leverage  as a function of Perp positions and assume that it's an accurate proxy for the entire token's leverage. In practice is this how things will play out? 

Is it possible for the magnitude of the spot and perp positions to drift apart from each other over time due to different protocol fee rates or differences in market prices?

What mechanism keeps the spot and short positions in proportion to each other and do they periodically need to be recalibrated?

Reply:
We are always buying/selling exact spot and virtual base tokens. So theoretically it isn't possible for the spot and perp positions to drift apart from each other due to different market prices.

There is no mechanism that keeps them in proportion. We set up a delta neutral position on engage, and then ensure that during each flow (rebalance/iterate rebalance/ripcord/reinvest) we add/subtract equal amounts of spot and perp position, using exact output/input trades.

Qq: How do different protocol fee rates affect the above logic?

Reply:
Ok `exact` makes sense, thanks for this explanation.